candlex/lib/candlex/backend.ex

```diff
@@ -0,0 +1,423 @@
+    tensor
+    |> Nx.BinaryBackend.constant(scalar, [])
+    |> Nx.BinaryBackend.backend_transfer(__MODULE__, backend_options)
+  end
+
+  @impl true
+  def from_binary(%T{shape: shape, type: type} = tensor, binary, _backend_options) do
+    # TODO: Don't ignore backend options
+
+    binary
+    |> Native.from_binary(to_candle_dtype(type), shape)
+    |> unwrap!()
+    |> to_nx(tensor)
+  end
+
+  @impl true
+  def iota(%T{shape: {}} = out, nil, backend_options) do
+    constant(out, 0, backend_options)
+  end
+
+  def iota(%T{shape: shape, type: type} = out, nil, _backend_options) do
+    Native.arange(0, Nx.size(shape), to_candle_dtype(type), shape)
+    |> unwrap!()
+    |> to_nx(out)
+  end
+
+  @impl true
+  def eye(%T{shape: shape, type: type} = out, backend_options) do
+    iota = Nx.iota(shape, backend_options)
+
+    Nx.equal(Nx.tril(iota), Nx.triu(iota))
+    |> Nx.as_type(type)
+  end
+
+  # Backend
+
+  @impl true
+  def backend_copy(%T{} = tensor, backend, backend_options) do
+    backend.from_binary(tensor, to_binary(tensor), backend_options)
+  end
+
+  @impl true
+  def backend_transfer(tensor, backend, backend_options) do
+    backend_copy(tensor, backend, backend_options)
+  after
+    backend_deallocate(tensor)
+  end
+
+  @impl true
+  def backend_deallocate(%T{} = _tensor) do
+    true
+  end
+
+  # Conversion
+
+  @impl true
+  def to_binary(tensor, _limit \\ nil) do
+    # TODO: don't ignore limit
+
+    from_nx(tensor)
+    |> Native.to_binary()
+    |> unwrap!()
+  end
+
+  # Aggregates
+
+  @impl true
+  def all(%T{} = out, %T{} = tensor, _opts) do
+    from_nx(tensor)
+    |> Native.all()
+    |> unwrap!()
+    |> to_nx(out)
+  end
+
+  # Element-wise
+
+  for op <- [:bitwise_and, :bitwise_or, :bitwise_xor, :left_shift, :right_shift] do
+    @impl true
+    def unquote(op)(out, l, r) do
+      # TODO: Implement in candle
+      out
+      |> Nx.BinaryBackend.unquote(op)(
+        backend_transfer(l, Nx.BinaryBackend, []),
+        backend_transfer(r, Nx.BinaryBackend, [])
+      )
+      |> backend_transfer(__MODULE__, [])
+    end
+  end
+
+  @impl true
+  def select(%T{shape: shape} = out, pred, on_true, on_false) do
+    on_true =
+      on_true
+      |> Nx.as_type(Nx.type(out))
+      |> from_nx()
+      |> Native.broadcast_to(shape)
+      |> unwrap!()
+
+    on_false =
+      on_false
+      |> Nx.as_type(Nx.type(out))
+      |> from_nx()
+      |> Native.broadcast_to(shape)
+      |> unwrap!()
+
+    pred
+    |> from_nx()
+    |> Native.where_cond(on_true, on_false)
+    |> unwrap!()
+    |> to_nx(out)
+  end
+
+  # Binary ops
+
+  for op <- [:add, :equal, :greater_equal, :less, :less_equal, :max, :min, :multiply, :subtract] do
+    @impl true
+    def unquote(op)(%T{} = out, %T{} = left, %T{} = right) do
+      {left, right} = maybe_upcast(left, right)
+      {left, right} = maybe_broadcast_bin_args(out.shape, left, right)
+
+      left
+      |> Native.unquote(op)(right)
```

Comment:
Could you use candle's broadcast_add instead?

Reply:
Oh, good call.

I'm already using the candle's `broadcast_*` variants for a few binary ops but I never removed broadcast here for those. Will do.

Reply:
Done in 9f7c26d8115d97322cd2624436c41e44b3104d46